engine.go

```diff
@@ -315,3 +334,104 @@ func (e *RTCEngine) readDataPacket(msg webrtc.DataChannelMessage) (*livekit.Data
 	err := proto.Unmarshal(msg.Data, dataPacket)
 	return dataPacket, err
 }
+
+func (e *RTCEngine) handleDisconnect() {
+	if e.closed.Load() {
+		return
+	}
+
+	if !e.reconnecting.CAS(false, true) {
+		return
+	}
+
+	go func() {
+		defer e.reconnecting.Store(false)
+		var reconnectCount int
+		var fullReconnect bool
+		for ; reconnectCount < maxReconnectCount; reconnectCount++ {
+			if fullReconnect {
+				if reconnectCount == 0 && e.OnRestarting != nil {
+					e.OnRestarting()
+				}
+				if err := e.restartConnection(); err != nil {
+					logger.Error(err, "restart connection failed")
+				} else {
+					return
+				}
+			} else {
+				if reconnectCount == 0 && e.OnResuming != nil {
+					e.OnResuming()
+				}
+				if err := e.resumeConnection(); err != nil {
+					logger.Error(err, "resume connection failed")
+					if !errors.Is(err, ErrSignalError) {
+						fullReconnect = true
+					}
+				} else {
+					return
+				}
+			}
+
+			delay := time.Duration(reconnectCount*reconnectCount) * initialReconnectInterval
+			if delay > maxReconnectInterval {
+				break
+			}
+			if reconnectCount < maxReconnectCount-1 {
+				time.Sleep(delay)
+			}
+		}
+
+		if e.OnDisconnected != nil {
+			e.OnDisconnected()
+		}
+	}()
+}
+
+func (e *RTCEngine) resumeConnection() error {
+	_, err := e.client.Join(e.url, e.token.Load(), e.connParams)
```

Comment:
I believe for resume, it needs to pass in `reconnect=1`, if it's not set, I don't think we would resume the session correctly.